### PR TITLE
TST: pytest 7.0.0, scipy 1.8.0 broke a few things

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is to fix some stuff that pytest 7.0.0 release broke.

Need new `asdf` release:

* asdf-format/asdf#1075 -- @WilliamJamieson 

Out of scope:

* #12820
* scipy/scipy#15545 (linkcheck)

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [ ] Do the proposed changes actually accomplish desired goals?
- [ ] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [ ] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [ ] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [ ] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [ ] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [ ] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [ ] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [ ] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by saimn** [comment]
Another issue is that `pytest.warns(None)` is deprecated, because "because many people used it to mean “this code does not emit warnings”, but it actually had the effect of checking that the code emits at least one warning of any type". 
```
❯ rg "pytest.warns\(None\)" | wc -l
35
```
We should replace those with the solutions advised here: https://docs.pytest.org/en/latest/how-to/capture-warnings.html#additional-use-cases-of-warnings-in-tests

**Comment by pllim** [comment]
Thanks, @saimn ! I'll get to that later today.

**Comment by dhomeier** [comment]
> We should replace those with the solutions advised here: https://docs.pytest.org/en/latest/how-to/capture-warnings.html#additional-use-cases-of-warnings-in-tests

For asdf @WilliamJamieson has fixed it in asdf-format/asdf#1075.

**Comment by dhomeier** [comment]
In addition there is an RTD warning/failure due to moved paths (along with the h5py one):
```
intersphinx inventory has moved: http://docs.h5py.org/en/stable/objects.inv -> https://docs.h5py.org/en/stable/objects.inv
intersphinx inventory has moved: https://pytest.readthedocs.io/en/stable/objects.inv -> https://docs.pytest.org/en/stable/objects.inv
```

File: astropy/io/fits/tests/test_header.py
Comment (pllim) [inline astropy/io/fits/tests/test_header.py:608]: Unless this is emitting a warning that we are purposely ignoring in `setup.cfg`, we do not need a check.

File: docs/conf.py
Comment (pllim) [inline docs/conf.py:101]: For the record, the URL in `sphinx-astropy` is already correct.

**Comment by pllim** [comment]
Hrrmmm... looks like `pytest.warns()` fails on zero warnings.

**Comment by pllim** [comment]
I am not sure if there is a good replacement for `pytest.warns(None)`. Or maybe `pytest.warns()` failing on zero warning is a bug upstream. 🤷

File: setup.cfg
Comment (pllim) [inline setup.cfg:83]: Tests are only passing with this pin, so that means we either have to:

* Pin `pytest<7`

*OR*

* Wait for `asdf` to release and then bump minversion of `asdf` to that release.

This pin broke RTD but that is okay because we will not merge with this pin anyway.

File: setup.cfg
Comment (pllim) [inline setup.cfg:83]: Is this too controversial for backport, @nden , @WilliamJamieson , or @eslavich ?

Comment (eslavich) [reply setup.cfg:83]: I don't think so, the worst thing someone would encounter upgrading from 2.6 --> 2.9.2 is a deprecation warning.  We've been pretty good about staying true to semantic versioning.

Comment (pllim) [reply setup.cfg:83]: Note: Looks like it is only tagged, not released yet.

File: astropy/cosmology/tests/test_funcs.py
Comment (pllim) [inline astropy/cosmology/tests/test_funcs.py:199]: Does this change matter?

```
        with pytest.warns(AstropyUserWarning, match=diag):
>           z0 = z_at_value(cosmo.angular_diameter_distance, 1*u.Gpc, zmax=2, maxfun=13, method=method)
E           Failed: DID NOT WARN. No warnings of type (<class 'astropy.utils.exceptions.AstropyUserWarning'>,) matching ('Solver returned None: Unsuccessful') were emitted. The list of emitted warnings is: [AstropyUserWarning('Solver returned None: \nMaximum number of iterations exceeded\nPrecision 0.023994025271917963 reached after 17 function calls.')].
```

**Comment by pllim** [comment]
OMG the CI is finally green. Only took me a few days... :rage4:

File: astropy/cosmology/tests/test_funcs.py
Comment (pllim) [inline astropy/cosmology/tests/test_funcs.py:204]: @nstarman , are you ok with this?

File: astropy/io/fits/tests/test_header.py
Comment (saimn) [reply astropy/io/fits/tests/test_header.py:608]: This probably comes from before the switch to use pytest's warning system.  But given that the aim of this test is to test that no warning is emitted, it probable makes sense to check that explicitly ? Or at least add a comment explaining that if a warning is emitted this would fail the test.

Comment (pllim) [reply astropy/io/fits/tests/test_header.py:608]: Okay, I added them back.

**Review by nstarman — COMMENTED** [review]
Cosmology section LGTM!

**Review by adrn — APPROVED** [review]
Coordinates change looks fine, if it works!

**Comment by pllim** [comment]
I see a few PR merges are getting held up by CI failures. So let's get this in. If the rest of the subpackage maintainers not happy with the diff, they can open follow-up PRs. Thanks!

**Comment by mhvk** [comment]
@pllim - thanks so much for what turned out to be quite the task! (And all great for units)

**Pull request merged**
